fleetdm/fleet
# Feature 5307: Add capability for parsing OVAL files (for Ubuntu)

Partially closes #5307 since this only adds supports for Ubuntu's OVAL files.
Closes #5313

We want to be able to ingests OVAL files and produce some sort of intermediate format that can be used for efficiently detect vulnerabilities. This PR adds the capability to both `fleet` and `fleetctl` to download and ingest OVAL files for the supported platforms (Ubuntu ATM).

The location of the OVAL definitions are contained in this [file](https://github.com/fleetdm/nvd/blob/master/oval_sources.json). We basically have two methods for syncing oval definitions:
- Manual sync: Launched when the user executes `fleetctl vulnerability-data-stream --dir some_dir`. Here we look at all hosts, collect a set of (platform, major(os_version)) and cross reference that with the `oval_sources.json` to figure out what OVAL definitions to download.
- Auto sync: Launched by our 'cron' runner. Here we use the `OsVersions stats` to efficiently determining what unique (platform, major(os_version)) exists and then we check the 'freshness' of the OVAL definitions that exist on disk (anything with a timestamp older than today will be deleted), we then use both criteria to figure out what new definitions will need to be downloaded.

# Testing instructions
1. Make sure fleetctl vulnerability-data-stream --dir some_dir downloads and parses oval files for all installed Ubuntu versions. The parsed files follow a `fleet_oval_ubuntu-<major_version>_<today's timestamp>.json` naming convention.
2. Make sure the oval definitions are automatically kept up to date by cron:
- Adding a new Ubuntu host (with a new version) should download a new OVAL definition.
- Removing an Ubuntu host should remove the OVAL definition (there will be delay here since we are using OS versions stats).
- Renaming an OVAL definition from `fleet_oval_ubuntu-<major_version>_<today's timestamp>.json` to `fleet_oval_ubuntu-<major_version>_<yesterday's timestamp>.json` should remove the file on the next cron tick.

# Checklist for submitter

If some of the following don't apply, delete the relevant line.

- [x] Changes file added for user-visible changes (in `changes/` and/or `orbit/changes/`).
